Source/WTF/wtf/JSONValues.cpp

```diff
@@ -646,38 +666,32 @@ void Value::writeJSON(StringBuilder& output) const
             output.append(m_value.number);
         break;
     }
-    default:
+    case Type::Object:
+    case Type::Array:
         ASSERT_NOT_REACHED();
     }
 }
 
 size_t Value::memoryCost() const
 {
-    size_t memoryCost = sizeof(this);
-    if (m_type == Type::String && m_value.string)
-        memoryCost += m_value.string->sizeInBytes();
-    return memoryCost;
+    return visitDerived([&](auto& derived) {
+        return derived.memoryCostImpl();
+    });
 }
 
-ObjectBase::~ObjectBase()
+size_t Value::memoryCostImpl() const
 {
+    size_t memoryCost = sizeof(*this);
+    if (m_type == Type::String && m_value.string)
+        memoryCost += m_value.string->sizeInBytes();
+    return memoryCost;
 }
 
-RefPtr<Object> ObjectBase::asObject()
-{
-    static_assert(sizeof(Object) == sizeof(ObjectBase), "cannot cast");
-    return static_cast<Object*>(this);
-}
-
-RefPtr<const Object> ObjectBase::asObject() const
-{
-    static_assert(sizeof(Object) == sizeof(ObjectBase), "cannot cast");
-    return static_cast<const Object*>(this);
-}
+ObjectBase::~ObjectBase() = default;
 
-size_t ObjectBase::memoryCost() const
+size_t ObjectBase::memoryCostImpl() const
 {
-    size_t memoryCost = Value::memoryCost();
+    size_t memoryCost = Value::memoryCostImpl();
```

Comment:
I thought about this more and I think we would be better off with `sizeof(*this)` here. Because calling `Value::memoryCostImpl` returns the wrong size, the size just of `Value`, and does a pointless check if it’s a string and we know it’s not. I don’t think we need a call through to the base class for future maintainability either.

Also I think that ideally we should include the memory used by the map itself, not just the memory used for the pointer to the hash table and for the keys and values, but also the memory used by the hash table. I’m honestly not sure how good our `memoryCost` functions needs to be to make garbage collection do the right thing. I suppose estimates are good enough. We could add a `memoryCost` function to HashMap but I suppose it would tricky to decide what to name it to make clear what it does and does not include.

Reply:
OK, changed to `sizeof(*this)`.